## Task-5

## Exploratory Data Analysis - Sports (Level - Advanced)

#### ● Perform ‘Exploratory Data Analysis’ on dataset ‘Indian Premier League’

### ● As a sports analysts, find out the most successful teams, players and factors contributing win or loss of a team.

### ● Suggest teams or players a company should endorse for its products. 

In [2]:
import numpy as np
import pandas as pd

In [3]:
deliveries_data = pd.read_csv('deliveries.csv')
matches_data = pd.read_csv('matches.csv')

In [4]:
matches_data.head(5)

id  season       city        date                        team1  \
0   1    2017  Hyderabad  2017-04-05          Sunrisers Hyderabad   
1   2    2017       Pune  2017-04-06               Mumbai Indians   
2   3    2017     Rajkot  2017-04-07                Gujarat Lions   
3   4    2017     Indore  2017-04-08       Rising Pune Supergiant   
4   5    2017  Bangalore  2017-04-08  Royal Challengers Bangalore   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   

          umpire1        umpire2 umpire3  
0     AY Dandekar       NJ Llong     NaN  
1  A Nand Kishore         S Ravi     NaN  
2     Nitin Menon      CK Nandan     NaN  
3    AK Chaudhary  C Shamshuddin     NaN  
4             NaN            NaN     NaN

In [5]:
matches_data.shape

(756, 18)

In [6]:
matches_data.describe()

id       season  dl_applied  win_by_runs  win_by_wickets
count    756.000000   756.000000  756.000000   756.000000      756.000000
mean    1792.178571  2013.444444    0.025132    13.283069        3.350529
std     3464.478148     3.366895    0.156630    23.471144        3.387963
min        1.000000  2008.000000    0.000000     0.000000        0.000000
25%      189.750000  2011.000000    0.000000     0.000000        0.000000
50%      378.500000  2013.000000    0.000000     0.000000        4.000000
75%      567.250000  2016.000000    0.000000    19.000000        6.000000
max    11415.000000  2019.000000    1.000000   146.000000       10.000000

In [302]:
matches_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 756 entries, 0 to 755
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   id               756 non-null    int64 
 1   season           756 non-null    int64 
 2   city             749 non-null    object
 3   date             756 non-null    object
 4   team1            756 non-null    object
 5   team2            756 non-null    object
 6   toss_winner      756 non-null    object
 7   toss_decision    756 non-null    object
 8   result           756 non-null    object
 9   dl_applied       756 non-null    int64 
 10  winner           752 non-null    object
 11  win_by_runs      756 non-null    int64 
 12  win_by_wickets   756 non-null    int64 
 13  player_of_match  752 non-null    object
 14  venue            756 non-null    object
 15  umpire1          754 non-null    object
 16  umpire2          754 non-null    object
 17  umpire3          119 non-null    ob

In [303]:
matche_drop_id = matches_data.drop(['id','date','dl_applied','umpire1','umpire2','umpire3','season','city'],axis=1)
matche_drop_id

team1                        team2  \
0            Sunrisers Hyderabad  Royal Challengers Bangalore   
1                 Mumbai Indians       Rising Pune Supergiant   
2                  Gujarat Lions        Kolkata Knight Riders   
3         Rising Pune Supergiant              Kings XI Punjab   
4    Royal Challengers Bangalore             Delhi Daredevils   
..                           ...                          ...   
751        Kolkata Knight Riders               Mumbai Indians   
752          Chennai Super Kings               Mumbai Indians   
753          Sunrisers Hyderabad               Delhi Capitals   
754               Delhi Capitals          Chennai Super Kings   
755               Mumbai Indians          Chennai Super Kings   

                     toss_winner toss_decision  result  \
0    Royal Challengers Bangalore         field  normal   
1         Rising Pune Supergiant         field  normal   
2          Kolkata Knight Riders         field  normal   
3                Kings XI Punjab         field  normal   
4    Royal Challengers Bangalore           bat  normal   
..                           ...           ...     ...   
751               Mumbai Indians         field  normal   
752          Chennai Super Kings           bat  normal   
753               Delhi Capitals         field  normal   
754          Chennai Super Kings         field  normal   
755               Mumbai Indians           bat  normal   

                          winner  win_by_runs  win_by_wickets player_of_match  \
0            Sunrisers Hyderabad           35               0    Yuvraj Singh   
1         Rising Pune Supergiant            0               7       SPD Smith   
2          Kolkata Knight Riders            0              10         CA Lynn   
3                Kings XI Punjab            0               6      GJ Maxwell   
4    Royal Challengers Bangalore           15               0       KM Jadhav   
..                           ...          ...             ...             ...   
751               Mumbai Indians            0               9       HH Pandya   
752               Mumbai Indians            0               6        AS Yadav   
753               Delhi Capitals            0               2         RR Pant   
754          Chennai Super Kings            0               6    F du Plessis   
755               Mumbai Indians            1               0       JJ Bumrah   

                                         venue  
0    Rajiv Gandhi International Stadium, Uppal  
1      Maharashtra Cricket Association Stadium  
2       Saurashtra Cricket Association Stadium  
3                       Holkar Cricket Stadium  
4                        M Chinnaswamy Stadium  
..                                         ...  
751                           Wankhede Stadium  
752                  M. A. Chidambaram Stadium  
753                           ACA-VDCA Stadium  
754                           ACA-VDCA Stadium  
755         Rajiv Gandhi Intl. Cricket Stadium  

[756 rows x 10 columns]

In [304]:
result = matche_drop_id['result'].replace('normal','won')
matche_cleaning = matche_drop_id.drop(['result'],axis=1)
matche_cleaning['result']= result
matche_cleaning.head()

team1                        team2  \
0          Sunrisers Hyderabad  Royal Challengers Bangalore   
1               Mumbai Indians       Rising Pune Supergiant   
2                Gujarat Lions        Kolkata Knight Riders   
3       Rising Pune Supergiant              Kings XI Punjab   
4  Royal Challengers Bangalore             Delhi Daredevils   

                   toss_winner toss_decision                       winner  \
0  Royal Challengers Bangalore         field          Sunrisers Hyderabad   
1       Rising Pune Supergiant         field       Rising Pune Supergiant   
2        Kolkata Knight Riders         field        Kolkata Knight Riders   
3              Kings XI Punjab         field              Kings XI Punjab   
4  Royal Challengers Bangalore           bat  Royal Challengers Bangalore   

   win_by_runs  win_by_wickets player_of_match  \
0           35               0    Yuvraj Singh   
1            0               7       SPD Smith   
2            0              10         CA Lynn   
3            0               6      GJ Maxwell   
4           15               0       KM Jadhav   

                                       venue result  
0  Rajiv Gandhi International Stadium, Uppal    won  
1    Maharashtra Cricket Association Stadium    won  
2     Saurashtra Cricket Association Stadium    won  
3                     Holkar Cricket Stadium    won  
4                      M Chinnaswamy Stadium    won

In [305]:
matche_cleaning.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 756 entries, 0 to 755
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   team1            756 non-null    object
 1   team2            756 non-null    object
 2   toss_winner      756 non-null    object
 3   toss_decision    756 non-null    object
 4   winner           752 non-null    object
 5   win_by_runs      756 non-null    int64 
 6   win_by_wickets   756 non-null    int64 
 7   player_of_match  752 non-null    object
 8   venue            756 non-null    object
 9   result           756 non-null    object
dtypes: int64(2), object(8)
memory usage: 59.2+ KB


In [306]:
matche_cleaned =matche_cleaning.dropna()

In [307]:
matche_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 752 entries, 0 to 755
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   team1            752 non-null    object
 1   team2            752 non-null    object
 2   toss_winner      752 non-null    object
 3   toss_decision    752 non-null    object
 4   winner           752 non-null    object
 5   win_by_runs      752 non-null    int64 
 6   win_by_wickets   752 non-null    int64 
 7   player_of_match  752 non-null    object
 8   venue            752 non-null    object
 9   result           752 non-null    object
dtypes: int64(2), object(8)
memory usage: 64.6+ KB


In [7]:
deliveries_data.head(5)

match_id  inning         batting_team                 bowling_team  over  \
0         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
1         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
2         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
3         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
4         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   

   ball    batsman non_striker    bowler  is_super_over  ...  bye_runs  \
0     1  DA Warner    S Dhawan  TS Mills              0  ...         0   
1     2  DA Warner    S Dhawan  TS Mills              0  ...         0   
2     3  DA Warner    S Dhawan  TS Mills              0  ...         0   
3     4  DA Warner    S Dhawan  TS Mills              0  ...         0   
4     5  DA Warner    S Dhawan  TS Mills              0  ...         0   

   legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
0            0            0             0             0           0   
1            0            0             0             0           0   
2            0            0             0             4           0   
3            0            0             0             0           0   
4            0            0             0             0           2   

   total_runs  player_dismissed dismissal_kind fielder  
0           0               NaN            NaN     NaN  
1           0               NaN            NaN     NaN  
2           4               NaN            NaN     NaN  
3           0               NaN            NaN     NaN  
4           2               NaN            NaN     NaN  

[5 rows x 21 columns]

In [8]:
deliveries_data.shape

(179078, 21)

In [10]:
deliveries_data.describe()

match_id         inning           over           ball  \
count  179078.000000  179078.000000  179078.000000  179078.000000   
mean     1802.252957       1.482952      10.162488       3.615587   
std      3472.322805       0.502074       5.677684       1.806966   
min         1.000000       1.000000       1.000000       1.000000   
25%       190.000000       1.000000       5.000000       2.000000   
50%       379.000000       1.000000      10.000000       4.000000   
75%       567.000000       2.000000      15.000000       5.000000   
max     11415.000000       5.000000      20.000000       9.000000   

       is_super_over      wide_runs       bye_runs    legbye_runs  \
count  179078.000000  179078.000000  179078.000000  179078.000000   
mean        0.000452       0.036721       0.004936       0.021136   
std         0.021263       0.251161       0.116480       0.194908   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       5.000000       4.000000       5.000000   

         noball_runs   penalty_runs   batsman_runs     extra_runs  \
count  179078.000000  179078.000000  179078.000000  179078.000000   
mean        0.004183       0.000056       1.246864       0.067032   
std         0.070492       0.016709       1.608270       0.342553   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       1.000000       0.000000   
75%         0.000000       0.000000       1.000000       0.000000   
max         5.000000       5.000000       7.000000       7.000000   

          total_runs  
count  179078.000000  
mean        1.313897  
std         1.605422  
min         0.000000  
25%         0.000000  
50%         1.000000  
75%         1.000000  
max        10.000000

In [9]:
deliveries_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 179078 entries, 0 to 179077
Data columns (total 21 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   match_id          179078 non-null  int64 
 1   inning            179078 non-null  int64 
 2   batting_team      179078 non-null  object
 3   bowling_team      179078 non-null  object
 4   over              179078 non-null  int64 
 5   ball              179078 non-null  int64 
 6   batsman           179078 non-null  object
 7   non_striker       179078 non-null  object
 8   bowler            179078 non-null  object
 9   is_super_over     179078 non-null  int64 
 10  wide_runs         179078 non-null  int64 
 11  bye_runs          179078 non-null  int64 
 12  legbye_runs       179078 non-null  int64 
 13  noball_runs       179078 non-null  int64 
 14  penalty_runs      179078 non-null  int64 
 15  batsman_runs      179078 non-null  int64 
 16  extra_runs        179078 non-null  int

In [311]:
deliveries_cleaned = deliveries_data.drop(['fielder','dismissal_kind','player_dismissed','bye_runs','legbye_runs','wide_runs','match_id',
                                            'batsman_runs','penalty_runs','noball_runs','is_super_over','inning'],axis=1)

In [312]:
deliveries_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 179078 entries, 0 to 179077
Data columns (total 9 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   batting_team  179078 non-null  object
 1   bowling_team  179078 non-null  object
 2   over          179078 non-null  int64 
 3   ball          179078 non-null  int64 
 4   batsman       179078 non-null  object
 5   non_striker   179078 non-null  object
 6   bowler        179078 non-null  object
 7   extra_runs    179078 non-null  int64 
 8   total_runs    179078 non-null  int64 
dtypes: int64(4), object(5)
memory usage: 12.3+ MB


##### As a sports analysts, find out the most successful teams, players .

In [313]:
matche_cleaned.head(2)

team1                        team2  \
0  Sunrisers Hyderabad  Royal Challengers Bangalore   
1       Mumbai Indians       Rising Pune Supergiant   

                   toss_winner toss_decision                  winner  \
0  Royal Challengers Bangalore         field     Sunrisers Hyderabad   
1       Rising Pune Supergiant         field  Rising Pune Supergiant   

   win_by_runs  win_by_wickets player_of_match  \
0           35               0    Yuvraj Singh   
1            0               7       SPD Smith   

                                       venue result  
0  Rajiv Gandhi International Stadium, Uppal    won  
1    Maharashtra Cricket Association Stadium    won

In [314]:
matche_cleaned.groupby(['result'])['winner'].value_counts()

result  winner                     
tie     Kings XI Punjab                  2
        Mumbai Indians                   2
        Rajasthan Royals                 2
        Delhi Capitals                   1
        Royal Challengers Bangalore      1
        Sunrisers Hyderabad              1
won     Mumbai Indians                 107
        Chennai Super Kings            100
        Kolkata Knight Riders           92
        Royal Challengers Bangalore     83
        Kings XI Punjab                 80
        Rajasthan Royals                73
        Delhi Daredevils                67
        Sunrisers Hyderabad             57
        Deccan Chargers                 29
        Gujarat Lions                   13
        Pune Warriors                   12
        Rising Pune Supergiant          10
        Delhi Capitals                   9
        Kochi Tuskers Kerala             6
        Rising Pune Supergiants          5
Name: winner, dtype: int64

In [315]:
matche_cleaned['player_of_match'].value_counts().sort_values(ascending=False)

CH Gayle          21
AB de Villiers    20
MS Dhoni          17
DA Warner         17
RG Sharma         17
                  ..
AP Tare            1
RE Levi            1
K Paul             1
DL Chahar          1
R Bhatia           1
Name: player_of_match, Length: 226, dtype: int64

### so from above we can say that Mumbai Indians is the best team and won by 107 times and the best player is CH Gayle

#### and now we will find the factors contributing win or loss of a team. 

In [316]:
matche_cleaned.columns

Index(['team1', 'team2', 'toss_winner', 'toss_decision', 'winner',
       'win_by_runs', 'win_by_wickets', 'player_of_match', 'venue', 'result'],
      dtype='object')

In [317]:
import statsmodels.api as sm

from sklearn.preprocessing import LabelEncoder
l= LabelEncoder()

team1=l.fit_transform(matche_cleaned["team1"])
team2=l.fit_transform(matche_cleaned["team2"])
toss_winner=l.fit_transform(matche_cleaned["toss_winner"])
toss_decision=l.fit_transform(matche_cleaned["toss_decision"])
winner=l.fit_transform(matche_cleaned["winner"])
player_of_match=l.fit_transform(matche_cleaned["player_of_match"])
venue=l.fit_transform(matche_cleaned["venue"])
result=l.fit_transform(matche_cleaned["result"])

matche_cleaned_new = matche_cleaned.drop(['team1','team2','toss_winner','toss_decision','winner',
                             'player_of_match','player_of_match','venue','result'],axis=1)

matche_cleaned_new['team1']=team1
matche_cleaned_new['team2']=team2
matche_cleaned_new['toss_winner']=toss_winner
matche_cleaned_new['toss_decision']=toss_decision
matche_cleaned_new['winner']=winner
matche_cleaned_new['player_of_match']=player_of_match
matche_cleaned_new['venue']=venue
matche_cleaned_new['result']=result

matche_cleaned_new.head(3)

x= matche_cleaned_new.drop(['result'],axis=1)
y=matche_cleaned_new["result"]
X_1 = sm.add_constant(x)
#Fitting sm.OLS model
model = sm.OLS(y,X_1).fit()
model.pvalues.round(2)

const              0.00
win_by_runs        0.00
win_by_wickets     0.00
team1              0.57
team2              0.50
toss_winner        0.52
toss_decision      0.83
winner             0.62
player_of_match    0.35
venue              0.26
dtype: float64

##### win_by_runs and win_by_wickets are  the factors contributing win or loss of a team. 

#### Now , we will find the best player in the given second dataset

In [319]:
deliveries_cleaned.head(2)

batting_team                 bowling_team  over  ball    batsman  \
0  Sunrisers Hyderabad  Royal Challengers Bangalore     1     1  DA Warner   
1  Sunrisers Hyderabad  Royal Challengers Bangalore     1     2  DA Warner   

  non_striker    bowler  extra_runs  total_runs  
0    S Dhawan  TS Mills           0           0  
1    S Dhawan  TS Mills           0           0

In [320]:
run = deliveries_cleaned['total_runs']-deliveries_cleaned['extra_runs']
deliveries_cleaned['run']=run
deliveries_cleaned.groupby(['batsman'])['run'].sum().sort_values(ascending=False)

batsman
V Kohli         5434
SK Raina        5415
RG Sharma       4914
DA Warner       4741
S Dhawan        4632
                ... 
J Denly            0
C Nanda            0
RR Bhatkal         0
K Ahmed            0
Abdur Razzak       0
Name: run, Length: 516, dtype: int64

###### here we can see that the best player is V Kohli  because he scored maximum run 5434

###### ● Suggest teams or players a company should endorse for its products. 

according to the both dataset , we can predict that 
Top best Team

Mumbai Inidans
Mumbai Indians                 
Chennai Super Kings            
Kolkata Knight Riders           
Royal Challengers Bangalore     
Kings XI Punjab 

and the top best players are  

V Kohli
SK Raina 
RG Sharma  
DA Warner 
S Dhawan  
CH Gayle          
AB de Villiers    
MS Dhoni          
DA Warner         
RG Sharma  

So, these above teams and players, company should endorse for its products

## Thankyou - End of Code